flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriterStateSerializer.java

```diff
@@ -59,14 +61,20 @@ public KafkaWriterState deserialize(int version, byte[] serialized) throws IOExc
         try (final ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
                 final DataInputStream in = new DataInputStream(bais)) {
             final String transactionalIdPrefix = in.readUTF();
+            int subtaskId = 0;
```

Comment:
Is this safe to do in case of a state migration now we have multiple states from subtaskId 0?

Reply:
It's a good point. I suggest to add a special variable UNKNOWN=-1 and disallow using POOLING then. In that way, we can guide the user to safely convert the checkpoint first to INCREMENTING and then POOLING.

Reply:
I added a new class TransactionOwnership that is used to document things but also to check the migration path.